### PR TITLE
Add option to enable code coverage support when building RPMs 

### PR DESCRIPTION
**Pull request by ArtiomDivak**
To enable building RPMs with code coverage support `-Dwith_coverage=1`
parameter can be passed to rpmbuild. Code coverage support is turned
off by default when building RPMs.

Related: https://github.com/containers/bluechi/issues/397
Signed-off-by: Artiom Divak <adivak@redhat.com>

### CODE REVIEW

**Comment by dougsland** [comment]
LGTM but I am wondering who will enable such feature in rpm?

**Comment by dougsland** [comment]
@ArtiomDivak in the description I see [WIP], is it still WIP? Looks good

File: bluechi.spec.in [hunk 1]
```diff
@@ -203,7 +208,11 @@ API description and manually written code to simplify recurring tasks.
 %autosetup
 
 %build
+%if %{with_coverage}
 %meson -Dapi_bus=system
+%else
+%meson -Dapi_bus=system -Dwith_coverage=true
+%endif
```
Comment (engelmi) [inline bluechi.spec.in]: Shouldn't this be reversed?
```suggestion
%if %{with_coverage}
%meson -Dapi_bus=system -Dwith_coverage=true
%else
%meson -Dapi_bus=system
%endif
```
If `with_coverage` is true, then add the meson config option.

File: README.developer.md [hunk 2]
```diff
@@ -333,6 +333,8 @@ The package build process can be adjusted using following environment variables:
   - To skip installation of build dependencies this option should contain `yes` value.
 - `WITH_PYTHON`
   - To skip building python bluechi modules this option should contain `0`.
+- `NO_COVERAGE`
+  - To start collecting coverage this option should contain `0`.
```
Comment (mwperina) [inline README.developer.md]: We don't want to build with code coverage by default, so using `WITH_COVERAGE` option makes more sense to me

File: bluechi.spec.in [hunk 3]
```diff
@@ -3,6 +3,10 @@
 %global with_python 1
 %endif
 
+# coverage collection is disabled by default , it can be enabled passing `--define "no_coverage 0"` option to rpmbuild
+%if 0%{!?no_coverage:1}
```
Comment (mwperina) [inline bluechi.spec.in]: Please use `with_coverage` instead

File: build-scripts/build-rpm.sh [hunk 4]
```diff
@@ -13,9 +13,10 @@ rpmbuild \
...
+    --define "no_coverage ${NO_COVERAGE:=1}"\
     --rebuild rpmbuild/SRPMS/*src.rpm
 
 # Move RPMs to exported artifacts
-ARTIFACTS_DIR="${ARTIFACTS_DIR:=$(pwd)/artifacts/rpms-$(date +%04Y%02m%02d%02H%02M)}"
-[[ -d $ARTIFACTS_DIR ]] || mkdir -p $ARTIFACTS_DIR
-find rpmbuild -iname \*rpm | xargs mv -t $ARTIFACTS_DIR
+# ARTIFACTS_DIR="${ARTIFACTS_DIR:=$(pwd)/artifacts/rpms-$(date +%04Y%02m%02d%02H%02M)}"
+# [[ -d $ARTIFACTS_DIR ]] || mkdir -p $ARTIFACTS_DIR
+# find rpmbuild -iname \*rpm | xargs mv -t $ARTIFACTS_DIR
```
Comment (mwperina) [inline build-scripts/build-rpm.sh]: What is this change for?

File: bluechi.spec.in [hunk 5]
```diff
@@ -3,6 +3,10 @@
...
+%global no_coverage 1
```
Comment (mwperina) [inline bluechi.spec.in]: I'd also define flags for meson here:

```
%global coverage_flags -Dwith_coverage=true
```

File: bluechi.spec.in [hunk 6]
```diff
@@ -203,7 +207,11 @@ API description and manually written code to simplify recurring tasks.
 %autosetup
 
 %build
+%if %{no_coverage}
 %meson -Dapi_bus=system
+%else
+%meson -Dapi_bus=system  -Dwith_coverage=true
+%endif
```
Comment (mwperina) [inline bluechi.spec.in]: Instead of above condition can't we use following code?

```
%build
%meson -Dapi_bus=system %{?coverage_flags}
```

**Review by mwperina — CHANGES_REQUESTED** [review]

File: bluechi.spec.in [hunk 7]
```diff
@@ -3,6 +3,10 @@
 %global with_python 1
 %endif
 
+# coverage collection is disabled by default , it can be enabled passing `--define "with_coverage 0"` option to rpmbuild
```
Comment (mwperina) [inline bluechi.spec.in]: To enable users need to pass `--define "with_coverage 1"`

File: build-scripts/build-rpm.sh (hunk 4 above)
Comment (mwperina) [reply build-scripts/build-rpm.sh]: Please revert, we need above commands to move artifacts into `$ARTIFACTS_DIR`, otherwise integration tests cannot install created RPMs

**Review by mwperina — CHANGES_REQUESTED** [review]

File: bluechi.spec.in [hunk 8]
```diff
@@ -3,6 +3,10 @@
 %global with_python 1
 %endif
 
+# coverage collection is disabled by default , it can be enabled passing `--define "with_coverage 1"` option to rpmbuild
+%if 0%{!?with_coverage:0}
+%global coverage_flags -Dwith_coverage=true
```
Comment (mwperina) [inline bluechi.spec.in:8]: Sorry, missed that before, but we need to use numbers as values for `with_coverage`:

```
%global coverage_flags -Dwith_coverage=0
```

**Review by mwperina — CHANGES_REQUESTED** [review]
Please also use more descriptive commit message, for example:

```
Add option to enable code coverage support when building RPMs

To enable building RPMs with code coverage support `-Dwith_coverage=1`
parameter can be passed to rpmbuild. Code coverage support is turned
off by default when building RPMs.

Related: https://github.com/containers/bluechi/issues/397
Signed-off-by: Artiom Divak <adivak@redhat.com>
```

File: bluechi.spec.in [hunk 9]
```diff
@@ -3,6 +3,10 @@
 %global with_python 1
 %endif
 
+# coverage collection is disabled by default , it can be enabled passing `--define "with_coverage 1"` option to rpmbuild
+%if 0%{!?with_coverage:0}
+%global coverage_flags -Dwith_coverage=0
+%endif
```
Comment (mwperina) [inline bluechi.spec.in:9]: Ah, sorry, only now realized that we could simplify this if-block to:

```
# Coverage collection is disabled by default , it can be enabled passing `--define "with_coverage 1"` option to rpmbuild
%global coverage_flags -Dwith_coverage=%{?with_coverage:0}
```

Comment (ArtiomDivak) [reply bluechi.spec.in:9]: But with_coverage variable is supposed to be true or false.  Should I change it in the meson to be 0 or 1 instead of true or false?

Comment (mwperina) [reply bluechi.spec.in:9]: Ah, forgot about it, so then we need to use longer code:

```
# Coverage collection is disabled by default , it can be enabled passing `--define "with_coverage 1"` option to rpmbuild
%if 0%{!?with_coverage:0}
%global coverage_flags -Dwith_coverage=true
%endif
```

**Review by mwperina — APPROVED** [review]
LGTM

File: test.txt
Comment (mwperina) [inline test.txt:1]: This file should not be part of the commit

**Review by mwperina — CHANGES_REQUESTED** [review]

File: bluechi.spec.in [hunk 10]
```diff
@@ -3,6 +3,10 @@
 %global with_python 1
 %endif
 
+# coverage collection is disabled by default , it can be enabled passing `--define "with_coverage 1"` option to rpmbuild
+%if 0%{?with_coverage}
```
Comment (mwperina) [inline bluechi.spec.in:7]: You should remove `?`, because it returns 1 if `with_coverage` exists, but not its value:

```
%if 0%{with_coverage}
%global coverage_flags -Dwith_coverage=true
%endif
```

File: bluechi.spec.in [hunk 11]
```diff
@@ -203,7 +207,7 @@ API description and manually written code to simplify recurring tasks.
 %autosetup
 
 %build
-%meson -Dapi_bus=system
+%meson -Dapi_bus=system %{?coverage_flags}
```
Comment (mwperina) [inline bluechi.spec.in:210]: Same here, you need to remove `?` and use just

```
%meson -Dapi_bus=system %{coverage_flags}
```

File: bluechi.spec.in (hunk 10 above)
Comment (mwperina) [reply bluechi.spec.in:7]: My mistake, the correct way should be:

```
%if 0%{?with_coverage}
%global coverage_flags -Dwith_coverage=true
%endif
```

File: bluechi.spec.in (hunk 11 above)
Comment (mwperina) [reply bluechi.spec.in:210]: My mistake, the correct way should be

```
%meson -Dapi_bus=system %{?coverage_flags}
```

**Review by mwperina — CHANGES_REQUESTED** [review]
Sorry, my mind is probably too tired today, so I gave you a bad advice. Please see my latest comments, they should be correct

**Comment by engelmi** [comment]
Please fetch the latest changes from `main` to get the Integration tests to pass.

**Review by mwperina — APPROVED** [review]
LGTM
